pytest-embedded/pytest_embedded/dut_factory.py

```diff
@@ -0,0 +1,682 @@
+    if 'openocd' not in _fixture_classes_and_options.classes:
+        return None
+
+    cls = _fixture_classes_and_options.classes['openocd']
+    kwargs = _fixture_classes_and_options.kwargs['openocd']
+    return cls(**_drop_none_kwargs(kwargs))
+
+
+def gdb_gn(_fixture_classes_and_options: ClassCliOptions) -> t.Optional['Gdb']:
+    if 'gdb' not in _fixture_classes_and_options.classes:
+        return None
+
+    cls = _fixture_classes_and_options.classes['gdb']
+    kwargs = _fixture_classes_and_options.kwargs['gdb']
+    return cls(**_drop_none_kwargs(kwargs))
+
+
+def qemu_gn(_fixture_classes_and_options: ClassCliOptions, app) -> t.Optional['Qemu']:
+    if 'qemu' not in _fixture_classes_and_options.classes:
+        return None
+
+    cls = _fixture_classes_and_options.classes['qemu']
+    kwargs = _fixture_classes_and_options.kwargs['qemu']
+
+    if 'app' in kwargs and kwargs['app'] is None:
+        kwargs['app'] = app
+
+    return cls(**_drop_none_kwargs(kwargs))
+
+
+def wokwi_gn(_fixture_classes_and_options: ClassCliOptions, app) -> t.Optional['WokwiCLI']:
+    """A wokwi subprocess that could read/redirect/write"""
+    if 'wokwi' not in _fixture_classes_and_options.classes:
+        return None
+
+    cls = _fixture_classes_and_options.classes['wokwi']
+    kwargs = _fixture_classes_and_options.kwargs['wokwi']
+
+    if 'app' in kwargs and kwargs['app'] is None:
+        kwargs['app'] = app
+    return cls(**_drop_none_kwargs(kwargs))
+
+
+def dut_gn(
+    _fixture_classes_and_options: ClassCliOptions,
+    openocd: t.Optional['OpenOcd'],
+    gdb: t.Optional['Gdb'],
+    app: App,
+    serial: t.Optional[t.Union['Serial', 'LinuxSerial']],
+    qemu: t.Optional['Qemu'],
+    wokwi: t.Optional['WokwiCLI'],
+) -> t.Union[Dut, t.List[Dut]]:
+    global dut_global_index
+    dut_global_index += 1
+
+    kwargs = _fixture_classes_and_options.kwargs['dut']
+    mixins = _fixture_classes_and_options.mixins['dut']
+
+    # since there's no way to know the target before setup finished
+    # we have to use the `app.target` to determine the dut class here
+    if hasattr(app, 'target') and app.target == 'linux':
+        from pytest_embedded_idf import LinuxDut
+
+        cls = LinuxDut
+        kwargs['serial'] = None  # replace it later with LinuxSerial
+    else:
+        cls = _fixture_classes_and_options.classes['dut']
+
+    for k, v in kwargs.items():
+        if v is None:
+            if k == 'app':
+                kwargs[k] = app
+            elif k == 'serial':
+                kwargs[k] = serial
+            elif k == 'openocd':
+                kwargs[k] = openocd
+            elif k == 'gdb':
+                kwargs[k] = gdb
+            elif k == 'qemu':
+                kwargs[k] = qemu
+            elif k == 'wokwi':
+                kwargs[k] = wokwi
+    return cls(**_drop_none_kwargs(kwargs), mixins=mixins)
+
+
+buff_parametrized_fixtures = {}
```

Comment:
For the first run, the buffer helps store some fixtures such as metadata, test case names, and services, etc., for reuse. It might be better to change the name to cache.

Reply:
Renamed to cache